### ISSUE TITLE
Patch to Time::Local

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4555]
Migrated from [rt.perl.org#7877](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7877) (status was 'resolved')

Searchable as RT7877$

**Comment by p5pRT** [comment]
### From ej_zg@hotmail.com



WHY doesn't perlbug send e\-mail correctly on Windows?

Summary&#8203;:
This call loops for 255 iterations and then fails\, instead of failing 
quickly&#8203;:
&nbsp;        Time&#8203;::Local&#8203;::timelocal\(01\, 00\, 16\, 31\, 11\, 69\) called at C&#8203;:/Program

Files/perl/site/lib/HTTP/Date\.pm line 55
&nbsp;        eval \{\.\.\.\} called at C&#8203;:/Program Files/perl/site/lib/HTTP/Date\.pm 
line 55
&nbsp;        HTTP&#8203;::Date&#8203;::str2time\('Wed Dec 31 16&#8203;:00&#8203;:01 1969'\) called at 
C&#8203;:/Program

Files/perl/site/lib/HTTP/Cookies\.pm line 277

The following patch fixes this\, plus makes Time&#8203;::Local "strict"\-safe\. It 
does not however

make "the right thing" happen\.

<details><summary>Inline Patch</summary>

```diff
--- Local.pm.orig	Wed Nov  7 00:52:24 2001
+++ Local.pm	Wed Nov  7 00:54:01 2001
@@ -1,25 +1,30 @@

```
</details>
package Time::Local;
require 5\.000;
\-require Exporter;
use Carp;
\+use strict;
\+use base qw\( Exporter \);

\-@&#8203;ISA        = qw\( Exporter \);
\+use vars qw\(@&#8203;EXPORT @&#8203;EXPORT\_OK\);
@&#8203;EXPORT        = qw\( timegm timelocal \);
@&#8203;EXPORT\_OK    = qw\( timegm\_nocheck timelocal\_nocheck \);

\# Set up constants
\-    $SEC  = 1;
\-    $MIN  = 60 \* $SEC;
\-    $HR   = 60 \* $MIN;
\-    $DAY  = 24 \* $HR;
\+    my $SEC  = 1;
\+    my $MIN  = 60 \* $SEC;
\+    my $HR   = 60 \* $MIN;
\+    my $DAY  = 24 \* $HR;
\# Determine breakpoint for rolling century
&nbsp;     my $thisYear = \(localtime\(\)\)\[5\];
\-    $nextCentury = int\($thisYear / 100\) \* 100;
\-    $breakpoint = \($thisYear \+ 50\) % 100;
\+    my $nextCentury = int\($thisYear / 100\) \* 100;
\+    my $breakpoint = \($thisYear \+ 50\) % 100;
&nbsp;     $nextCentury \+= 100 if $breakpoint \< 50;

my %options;

\+\# aurgh\, globals\!
\+my %cheat;
\+my $ym;
\+
sub timegm \{
&nbsp;     my \(@&#8203;date\) = @&#8203;\_;
&nbsp;     if \($date\[5\] > 999\) \{
@&#8203;@&#8203; \-29\,8 \+34\,8 @&#8203;@&#8203;
&nbsp;         $date\[5\] \-= 100 if $date\[5\] > $breakpoint;
&nbsp;         $date\[5\] \+= $nextCentury;
&nbsp;     \}
\-    $ym = pack\(C2\, @&#8203;date\[5\,4\]\);
\-    $cheat = $cheat\{$ym\} || &cheat\(@&#8203;date\);
\+    $ym = pack\('C2'\, @&#8203;date\[5\,4\]\);
\+    my $cheat = $cheat\{$ym\} || &cheat\(@&#8203;date\);
&nbsp;     $cheat
&nbsp;     \+ $date\[0\] \* $SEC
&nbsp;     \+ $date\[1\] \* $MIN
@&#8203;@&#8203; \-71\,8 \+76\,8 @&#8203;@&#8203;

&nbsp;     $tzsec \+= $HR if\($lt\[8\]\);

\-    $time = $t \+ $tzsec;
\-    @&#8203;test = localtime\($time \+ \($tt \- $t\)\);
\+    my $time = $t \+ $tzsec;
\+    my @&#8203;test = localtime\($time \+ \($tt \- $t\)\);
&nbsp;     $time \-= $HR if $test\[2\] \!= $\_\[2\];
&nbsp;     $time;
\}
@&#8203;@&#8203; \-83\,8 \+88\,8 @&#8203;@&#8203;
\}

sub cheat \{
\-    $year = $\_\[5\];
\-    $month = $\_\[4\];
\+    my $year = $\_\[5\];
\+    my $month = $\_\[4\];
&nbsp;     unless \($options\{no\_range\_check\}\) \{
&nbsp;    croak "Month '$month' out of range 0\.\.11" if $month > 11 || $month \< 0;
&nbsp;    croak "Day '$\_\[3\]' out of range 1\.\.31"      if $\_\[3\] > 31 || $\_\[3\] \< 1;
@&#8203;@&#8203; \-92\,21 \+97\,23 @&#8203;@&#8203;
&nbsp;    croak "Minute '$\_\[1\]' out of range 0\.\.59" if $\_\[1\] > 59 || $\_\[1\] \< 0;
&nbsp;    croak "Second '$\_\[0\]' out of range 0\.\.59" if $\_\[0\] > 59 || $\_\[0\] \< 0;
&nbsp;     \}
\-    $guess = $^T;
\-    @&#8203;g = gmtime\($guess\);
\-    $lastguess = "";
\-    $counter = 0;
\-    while \($diff = $year \- $g\[5\]\) \{
\+    my $guess = $^T;
\+    my @&#8203;g = gmtime\($guess\);
\+    my $lastguess = "";
\+    my $counter = 0;
\+    my $thisguess;
\+    while \(my $diff = $year \- $g\[5\]\) \{
&nbsp;    croak "Can't handle date \("\.join\("\, "\,@&#8203;\_\)\."\)" if \+\+$counter > 255;
&nbsp;    $guess \+= $diff \* \(363 \* $DAY\);
\-    @&#8203;g = gmtime\($guess\);
\+    croak "Can't handle date \("\.join\("\, "\,@&#8203;\_\)\."\)" unless @&#8203;g = gmtime\($guess\);
&nbsp;    if \(\($thisguess = "@&#8203;g"\) eq $lastguess\)\{
&nbsp;        croak "Can't handle date \("\.join\("\, "\,@&#8203;\_\)\."\)";
&nbsp;        \#date beyond this machine's integer limit
&nbsp;    \}
&nbsp;    $lastguess = $thisguess;
&nbsp;     \}
\-    while \($diff = $month \- $g\[4\]\) \{
\+
\+    while \(my $diff = $month \- $g\[4\]\) \{
&nbsp;    croak "Can't handle date \("\.join\("\, "\,@&#8203;\_\)\."\)" if \+\+$counter > 255;
&nbsp;    $guess \+= $diff \* \(27 \* $DAY\);
&nbsp;    @&#8203;g = gmtime\($guess\);
@&#8203;@&#8203; \-116\,7 \+123\,7 @&#8203;@&#8203;
&nbsp;    \}
&nbsp;    $lastguess = $thisguess;
&nbsp;     \}
\-    @&#8203;gfake = gmtime\($guess\-1\); \#still being sceptic
\+    my @&#8203;gfake = gmtime\($guess\-1\); \#still being sceptic
&nbsp;     if \("@&#8203;gfake" eq $lastguess\)\{
&nbsp;         croak "Can't handle date \("\.join\("\, "\,@&#8203;\_\)\."\)";
&nbsp;         \#date beyond this machine's integer limit

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.6.1:

Configured by Administrator at Mon Aug 20 12:26:50 2001.

Summary of my perl5 (revision 5 version 6 subversion 1) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define 
usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cl', ccflags ='-nologo -O1 -MD -DNDEBUG -DWIN32 -D_CONSOLE 
-DNO_STRICT

-DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS 
-DPERL_MSVCRT_READFIX',
    optimize='-O1 -MD -DNDEBUG',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=4
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release  
-libpath:"C:\Program

Files\perl\lib\CORE"  -machine:x86'
    libpth="C:\Program Files\Microsoft Visual Studio\VC98\mfc\lib" 
"C:\Program

Files\Microsoft Visual Studio\VC98\lib" "C:\Program Files\perl\lib\CORE"
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib

advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib 
wsock32.lib mpr.lib

winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib

advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib 
wsock32.lib mpr.lib

winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl56.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release  
-libpath:"C:\Program

Files\perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.1:
    C:/Program Files/perl/lib
    C:/Program Files/perl/site/lib
    .


Environment for perl v5.6.1:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Program 
Files\perl\bin\;C:\WINNT\system32;C:\WINNT;C:\WINNT\System32\Wbem
    PERL_BADLANG (unset)
    SHELL (unset)


_________________________________________________________________
Get your FREE download of MSN Explorer at http://explorer.msn.com/intl.asp

```
</details>

**Comment by p5pRT** [comment]
@cwest - Status changed from 'open' to 'resolved'
